Implement `MaybeDangling` compiler support

WaffleLapkin opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/150447)*
<!-- homu-ignore:end -->

Tracking issue: https://github.com/rust-lang/rust/issues/118166

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

cc @RalfJung

WaffleLapkin committed:
add an adt flag for `MaybeDangling`

WaffleLapkin committed:
add semantics to `MaybeDangling`

WaffleLapkin committed:
add a codegen-llvm test for `ManuallyDrop`

RalfJung commented on compiler/rustc_middle/src/ty/layout.rs:
```diff
@@ -1053,6 +1053,12 @@ where
+                // FIXME: what is the exact effect of maybe dangling?
+                Self::ty_and_layout_pointee_info_at(this.field(cx, 0), cx, offset)
+                    .map(|info| PointeeInfo { safe: None, ..info })
+            }
```
This should also reset the `size` to 0, that's how we encode `dereferenceable`.

Odd that the test does not catch this...

WaffleLapkin committed:
refactor `PointeeInfo`

Make `size`/`align` always correct rather than conditionally on the
`safe` field. This makes it less error prone and easier to work with for
`MaybeDangling` / potential future pointer kinds like `Aligned<_>`.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/22194040886/job/64189580259) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/64189580259) [(plain)](https://github.com/rust-lang/rust/commit/6d96bdbaab1cd002d23da68b919654a3674ab720/checks/64189580259/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::ToolBuild { build_compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 13.757
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
Diff in /checkout/compiler/rustc_codegen_ssa/src/mir/statement.rs:74:
             mir::StatementKind::Intrinsic(box NonDivergingIntrinsic::CopyNonOverlapping(
                 mir::CopyNonOverlapping { ref count, ref src, ref dst },
             )) => {
-
                 let dst_val = self.codegen_operand(bx, dst);
                 let src_val = self.codegen_operand(bx, src);
                 let count = self.codegen_operand(bx, count).immediate();
Diff in /checkout/compiler/rustc_codegen_ssa/src/mir/statement.rs:81:
 
-                let &ty::RawPtr(pointee, _) = dst_val.layout.ty.kind() else { bug!("expected pointer") };
-                let pointee_layout= bx.tcx().layout_of(bx.typing_env().as_query_input(pointee)).expect("expected pointee to have a layout");
+                let &ty::RawPtr(pointee, _) = dst_val.layout.ty.kind() else {
+                    bug!("expected pointer")
+                };
+                let pointee_layout = bx
+                    .tcx()
+                    .layout_of(bx.typing_env().as_query_input(pointee))
+                    .expect("expected pointee to have a layout");
 
                 let bytes = bx.mul(count, bx.const_usize(pointee_layout.layout.size().bytes()));
 
fmt: checked 6707 files
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:00:47
  local time: Thu Feb 19 18:13:27 UTC 2026
```

</details>

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/22194946717/job/64192764278) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/64192764278) [(plain)](https://github.com/rust-lang/rust/commit/d71787d081fbbb18e8890f8c29820cde9c1d361f/checks/64192764278/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

=> Removing the following docker images:
WARNING: This output is designed for human readability. For machine-readable output, please use --format.
IMAGE                                               ID             DISK USAGE   CONTENT SIZE   EXTRA
ghcr.io/dependabot/dependabot-updater-core:latest   e5f6a669cd91       1.18GB          310MB        
=> Removing docker images...
Deleted Images:
untagged: ghcr.io/dependabot/dependabot-updater-core:latest
deleted: sha256:e5f6a669cd91709f67772f14e1a32ab56ea12631fb213c32f8396339ac68e07b
deleted: sha256:a6e51fba71b953db04611aa17bc8045847233f0d64d7f1dd56b778d777b9a0e6
deleted: sha256:c30f9e1abc0878de2082c6acb3a9257a516e0d6060cbe7fa29405a658b29aaca

Total reclaimed space: 309.7MB
---
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error[E0308]: mismatched types
   --> compiler/rustc_codegen_gcc/src/type_of.rs:292:47
    |
292 |                     cx.type_pointee_for_align(pointee.align)
    |                        ---------------------- ^^^^^^^^^^^^^ expected `Align`, found `Option<Align>`
    |                        |
    |                        arguments to this method are incorrect
    |
    = note: expected struct `rustc_abi::Align`
                 found enum `std::option::Option<rustc_abi::Align>`
note: method defined here
   --> compiler/rustc_codegen_gcc/src/type_.rs:81:12
    |
 81 |     pub fn type_pointee_for_align(&self, align: Align) -> Type<'gcc> {
    |            ^^^^^^^^^^^^^^^^^^^^^^        ------------
help: consider using `Option::expect` to unwrap the `std::option::Option<rustc_abi::Align>` value, panicking if the value is an `Option::None`
    |
292 |                     cx.type_pointee_for_align(pointee.align.expect("REASON"))
    |                                                            +++++++++++++++++

For more information about this error, try `rustc --explain E0308`.
[RUSTC-TIMING] rustc_codegen_gcc test:false 1.174
error: could not compile `rustc_codegen_gcc` (lib) due to 1 previous error
```

</details>

WaffleLapkin committed:
make `PointeeInfo::align` non-optional

Instead of defaulting to `None` it now defaults to `Align::ONE` i.e.
no alignment restriction. Codegen test changes are due to us now skipping
`align 1` annotations (they are useless; not skipping them makes all the
raw pointers gain an `align 1` annotation which doesn't seem any good)

RalfJung commented on compiler/rustc_middle/src/ty/layout.rs:
```diff
@@ -1023,40 +1023,88 @@ where
+
+                tcx.layout_of(typing_env.as_query_input(pointee)).ok().map(|layout| PointeeInfo {
+                    safe: Some(PointerKind::Box {
                         unpin: optimize
```
```suggestion
                        // Same logic as for mutable references above.
                        unpin: optimize
```

RalfJung reviewed (changes requested):
This looks generally great, thanks! Nice cleanup for getting rid of those `align 1` everywhere.

I left some comments. I don't understand why the `may_dangle` field exists. Please resolve the conflicts so we can do a perf run.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150447/69b78537fac74de40f009b076bcbbf54b77683ad..efc351d51181f01ade8ac567a98fccac55701246)*

RalfJung commented:
Indeed, that is the *intent* of this PR. ;)

Note that this PR does not *guarantee* the absence of those optimizations on stable. It only guarantees it for code that uses `MaybeDangling` directly, which remains unstable.

WaffleLapkin committed:
strip readonly/captures from `MaybeDangling<&T>`

meithecatte commented on tests/codegen-llvm/manually_drop_refs.rs:
Are you sure that `MaybeDangling<&T>` actually need to strip `captures` and `readonly`? Those do not imply any validity invariants, and effectively only cause restrictions on any `&T`s that get extracted out of the `MaybeDangling` and how *they* can be used.

That being said, I do not know what the miri paragraph in the RFC is saying because I don't know off the top of my head what retagging is, so... sorry if that's related to the answer to my question (though, even then I feel like this should be spelled out more explicitly).

WaffleLapkin commented:
@bors try @rust-timer queue

RalfJung reviewed (approved):
I like the new test, thanks. :)

Minor nit on the doc comment. IIUC the scalar is only treated like `&mut` if it's a ptr scalar; if it is an int, `drop_target_pointee` gets ignored.

r=me with that fixed. Thanks *so much* for finally pushing this forward ❤️ 

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150447/1b7d722f429f09c87b08b757d89c689c6cf7f6e7..d5e1b0e14a32dbb3f33f997adfa2c621d7725fb0)*

rust-timer commented:
Finished benchmarking commit (64b72a1fa5449d928d5f553b01a596b78ee255d2): [comparison URL](https://perf.rust-lang.org/compare.html?start=70d86e3abeecf3a655264d9a716c5d08160176b7&end=64b72a1fa5449d928d5f553b01a596b78ee255d2&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=70d86e3abeecf3a655264d9a716c5d08160176b7&end=64b72a1fa5449d928d5f553b01a596b78ee255d2&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.5% | [-0.5%, -0.5%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=70d86e3abeecf3a655264d9a716c5d08160176b7&end=64b72a1fa5449d928d5f553b01a596b78ee255d2&stat=max-rss)
<details>
<summary>Results (primary 0.3%, secondary 3.5%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.7%  | [0.8%, 2.7%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 3.5%  | [3.4%, 3.6%]   | 2     |
| Improvements ✅ <br /> (primary)   | -2.5% | [-2.5%, -2.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.3%  | [-2.5%, 2.7%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=70d86e3abeecf3a655264d9a716c5d08160176b7&end=64b72a1fa5449d928d5f553b01a596b78ee255d2&stat=cycles:u)
<details>
<summary>Results (primary -1.3%, secondary -4.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.3% | [-1.3%, -1.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -4.0% | [-4.0%, -4.0%] | 1     |
| All ❌✅ (primary)                 | -1.3% | [-1.3%, -1.3%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=70d86e3abeecf3a655264d9a716c5d08160176b7&end=64b72a1fa5449d928d5f553b01a596b78ee255d2&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary -0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.0% | [-0.0%, -0.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 2     |
| All ❌✅ (primary)                 | -0.0% | [-0.0%, -0.0%] | 1     |
</details>

**Bootstrap**: 480.787s -> 479.735s (-0.22%)
**Artifact size**: 395.02 MiB -> 395.00 MiB (-0.00%)
